### PR TITLE
Pass tools as None to `apply_chat_template` when it is an empty list

### LINKED ISSUE #5379: GRPO Trainer: Empty tools not handled properly while tokenizing the prompts.
### Reproduction

TRL GRPO Trainer tokenizes the prompts beforehand using this code.

```python
tokenized = self.processing_class.apply_chat_template(
    conversation=prompts,
    tools=self.tools,
    chat_template=self.chat_template,
    add_generation_prompt=True,
    tokenize=True,
    return_dict=True,
    padding=True,
    **self.chat_template_kwargs,
)
```

When no tools are passed, `self.tools` is an empty list. Some models like `meta-llama/Llama-3.1-8B-Instruct` just check if `tools` is defined (i.e., tools is not None).  This creates a difference in the processed format when tools is None and an empty list. This results in a sudden change in the distribution of text when some model is first finetuned using `SFTTrainer` and then  on `GRPOTrainer`

The following is the output of `tokenizer.apply_chat_template` when tools is not passed or passed as None.
```
<|begin_of_text|><|start_header_id|>system<|end_header_id|>

Cutting Knowledge Date: December 2023
Today Date: 26 Jul 2024

<|eot_id|><|start_header_id|>user<|end_header_id|>

Tell me a joke<|eot_id|><|start_header_id|>assistant<|end_header_id|>
``` 

The following is the output of `tokenizer.apply_chat_template` when tools is not passed as an empty list.
```
<|begin_of_text|><|start_header_id|>system<|end_header_id|>

Environment: ipython
Cutting Knowledge Date: December 2023
Today Date: 26 Jul 2024

<|eot_id|><|start_header_id|>user<|end_header_id|>

Given the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.

Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}.Do not use variables.

Tell me a joke<|eot_id|><|start_header_id|>assistant<|end_header_id|>
```

The solution would be to pass None as `tools` when `self.tools` is empty.

### System Info

- Platform: Linux-6.18.5.1.amd64-smp-x86_64-with-glibc2.41
- Python version: 3.13.5
- TRL version: 1.0.0rc1
- PyTorch version: 2.9.1+cu128
- accelerator(s): cpu
- Transformers version: 4.57.6
- Accelerate version: 1.13.0
- Accelerate config:
  - compute_environment: LOCAL_MACHINE
  - distributed_type: DEEPSPEED
  - mixed_precision: bf16
  - use_cpu: False
  - debug: False
  - num_processes: 2
  - machine_rank: 0
  - num_machines: 1
  - rdzv_backend: static
  - same_network: True
  - main_training_function: main
  - enable_cpu_affinity: False
  - deepspeed_config: {'gradient_accumulation_steps': 1, 'offload_optimizer_device': 'none', 'offload_param_device': 'none', 'zero3_init_flag': False, 'zero_stage': 2}
  - downcast_bf16: no
  - tpu_use_cluster: False
  - tpu_use_sudo: False
  - tpu_env: []
- Datasets version: 4.8.3
- HF Hub version: 0.36.2
- bitsandbytes version: 0.49.2
- DeepSpeed version: 0.18.8
- Liger-Kernel version: 0.7.0
- LLM-Blender version: not installed
- OpenAI version: 2.24.0
- PEFT version: 0.18.1
- vLLM version: 0.16.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by rabinadk1**
Fixes #5379

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Low Risk**
> Low risk: this is a small, consistent parameter normalization that only affects how chat templates are rendered when `tools` is an empty list.
> 
> **Overview**
> Fixes tool-calling prompt rendering by normalizing `tools=[]` to `tools=None` before calling `apply_chat_template`, working around a Llama template bug that injects unwanted tool boilerplate.
> 
> This behavior is applied consistently across dataset templating (`data_utils.apply_chat_template`) and GRPO/DPPO/Async GRPO rollout/tokenization paths so empty tool lists no longer alter the rendered prompt/ids.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 68c6f445b351ea787fedd3601e72b1e90d378d0b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1261,7 +1261,7 @@ def _tokenize_prompts(self, prompts: list):
             # See: https://github.com/huggingface/transformers/issues/44514
             tokenized = self.processing_class.apply_chat_template(
                 conversation=prompts,
-                tools=self.tools,
+                tools=self.tools or None,
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: Good point, can you add the same modif everywhere we call apply_chat_template with tools?

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1261,7 +1261,7 @@ def _tokenize_prompts(self, prompts: list):
...
                 chat_template=self.chat_template,
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:1265]: I'm not sure it's absolutely necessary

File: trl/trainer/grpo_trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rabinadk1) [reply trl/trainer/grpo_trainer.py]: Done!

**Comment by qgallouedec** [comment]
@codex review

**Comment by qgallouedec** [comment]
thanks! I just added a comment so that we remember why we have to pass `or None`

**Comment by qgallouedec** [comment]
CI is failing for experimental but it's not related

**Comment by qgallouedec** [comment]
CI is failing for the distributed smoke test, it doesn't come from this PR, most likely from transformers v5.4.0, I'll merge this PR
